File: credit.c
#include <stdio.h>
#include <cs50.h>

int digsum(long);

int main(void)
{

    long long number = get_long("Number: ");

    int sum1 = 0, sum2 = 0, total;
    int num_dig = 0;
    int i = 0;
    int comp = -1;

    int fin = 0;
    if (number < 1000000000000)
    {
        printf("INVALID\n");
        fin = 1;
    }


     while(number != 0)
     {
         if(number/100 == 0 && comp == -1)
         {

             if(number > 10)
             {
                 if (number == 34 || number == 37)
                 {
                     comp = 1; //Amex
                 }

                 else if(number == 51 || number == 52 || number == 53 || number == 54 || number == 55)
                 {
                     comp = 2; // MasterCard
                 }
             }

             if (number < 10)
             {
                 if (number == 4)
                 {

                     comp = 3; //visa
                 }

                 else
                 {
                     comp = 4 ; //Invalid
                 }
             }
         }
         if (i % 2 == 0)
         {
             sum1 += digsum(number % 10);

         }
         else
         {
             sum2 += digsum((number % 10 ) * 2);

         }
         i++;
         number = number / 10;
     }
     total = sum1 + sum2;
     if (total % 10 == 0 && !fin)
     {
         switch (comp)
         {
             case 1:
             printf("AMEX\n");
             break;

             case 2:
             printf("MASTERCARD\n");
             break;

             case 3:
             printf("VISA\n");
             break;

             case 4:
             printf("INVALID\n");
             break;
         }
     }

     else if (!fin)
     {
         printf("INVALID\n");
     }
    }

int digsum(long num)
{
    int sum = 0;
    num = (int) num;

    while(num>0)
    {
        sum += num % 10;
        num /= 10;
    }
    return sum;
}

File: credit.py
from cs50 import *
n = get_int("Number: ")
print()
#print(f"n = {n}")
first_digits = n % 100
sum1 = sum2 = i = 0

while n != 0:
    if i % 2 == 0:
        sum2 += n % 10
        #print(f"sum2 now is: {sum2}")
    else:
        sum1 += (2 * (n % 10))%10 + (2 * (n % 10))//10
        #print(f"sum1 now is: {sum1}")
    i += 1
    n = n // 10

Luhn_valid = (sum1+sum2) % 10 == 0
#print(f"sum1 = {sum1}, sum2 = {sum2}")
if Luhn_valid:
    if first_digits % 10 == 4 and (i == 13 or i == 16):
        print("VISA")
    elif first_digits >= 51 and first_digits <= 55 and i == 16:
        print("MASTERCARD")
    elif first_digits == 34 or first_digits == 37 and i == 15:
        print("AMEX")
    else:
        print("INVALID")
else:
    print("INVALID")
